### PR TITLE
Neclab/ngsi introduction

### PR DESCRIPTION
**Pull request by tobiasjacobs**
This is my first pull request, so please apologize any abuse of conventions.

I believe that also the new NGSI version needs an complete description of the main concepts of entities, attributes, etc. and I wrote an introduction section accordingly. 

It is not yet complete, as also subscriptions, registrations, etc. will be needed, but I like to separate this out to another pull request.

### CODE REVIEW

File: doc/apiary/v2/fiware-ngsiv2-reference.apib [hunk 1]
```diff
@@ -21,6 +21,47 @@ This specification is licensed under the
...
 (http://forge.fiware.org/plugins/mediawiki/wiki/fiware/index.php/FI-WARE_Open_Specification_Legal_Notice_%28essential_patents_license%29).
 
+## FIWARE NGSI information and data model introduction
+
+The FIWARE NGSI (Next Generation Service Interface) API defines 
+
+* a **data model** for context information, based on a simple information model using the notion of *context entities*.
+* a **context data interface** for exchanging information by means of query, subscription, and update operations.
+* a **context availability interface** for exchanging information on how to obtain context information.
+
```
Comment (kzangeli) [inline doc/apiary/v2/fiware-ngsiv2-reference.apib]: If you end the bullet points with a dot, you should probably start then with uppercase letter.
So, either remove the dots or start with uppercase.

File: doc/apiary/v2/fiware-ngsiv2-reference.apib [hunk 2]
```diff
@@ -21,6 +21,47 @@ This specification is licensed under the
...
+Context entities, or simply called entities, are the center of gravity in the FIWARE NGSI information model. An entity represents a real-world object like e.g. a sensor, a person, or a room. Each entity is uniquely identified by its **entity id**. 
+
+Furthermore, the type system of FIWARE NGSI enables entities to have an **entity type**. Entity types are semantic types; they are intended to describe the type of real-world object represented by the entity. For example, a context entity with id *sensor-365* could have the type *temperatureSensor*.
+
+### Context Elements
+
+While context entities are the central concept of the FIWARE NGSI information model, context elements are the corresponding type of data object in the NGSI *data model*. 
+
+A context element is a JSON object  which contains information about a specific context entity. Consequently, a context element has a mandatory property 'id', an optional property 'type', and, optionally, further properties which describe more information about the entity (see the **context attributes** section below). Details of the JSON representation of context elements can be found below.
```
Comment (kzangeli) [inline doc/apiary/v2/fiware-ngsiv2-reference.apib]: I wouldn't say a context element is a JSON object.
Rather just an "object that can be represented as a JSON object".
Double space between 'object' and 'which'

File: doc/apiary/v2/fiware-ngsiv2-reference.apib [hunk 3]
```diff
@@ -21,6 +21,47 @@ This specification is licensed under the
...
+
+It is important to understand that the relationship of entity and context element is one-to-many. This means that 
+* each context element refers to exactly one entity
+* there can be several context elements  referring to the same entity. The context elements can for example contain different pieces of information about the entity.
+
+### Context Attributes 
+
+Context attributes are properties of context entities. For example, the current speed of a car could be modeled as attribute *current_speed* of entity *car-104*.
+
+In the NGSI data model, attributes have an *attribute name*, and *attribute type*, an *attribute value* and *metadata*.
```
Comment (kzangeli) [inline doc/apiary/v2/fiware-ngsiv2-reference.apib]: Typo: and -> an    (In 'attribute name, *and* attribute type')

File: doc/apiary/v2/fiware-ngsiv2-reference.apib [hunk 4]
```diff
@@ -21,6 +21,47 @@ This specification is licensed under the
...
+* each context element refers to exactly one entity
+* there can be several context elements  referring to the same entity. The context elements can for example contain different pieces of information about the entity.
+
+### Context Attributes 
+
+Context attributes are properties of context entities. For example, the current speed of a car could be modeled as attribute *current_speed* of entity *car-104*.
+
+In the NGSI data model, attributes have an *attribute name*, and *attribute type*, an *attribute value* and *metadata*. 
+ * the attribute name describes which kind of property of the entity the attribute value represents, for example *current_speed*.
+ * the attribute type represents the NGSI value type of the attribute value. Note that FIWARE NGSI has an own type system for attribute values, so NGSI value types are not the same as JSON types.
```
Comment (kzangeli) [inline doc/apiary/v2/fiware-ngsiv2-reference.apib]: an own type => its own type   ("an own type" sounds a little funny, right?)

File: doc/apiary/v2/fiware-ngsiv2-reference.apib [hunk 5]
```diff
@@ -21,6 +21,47 @@ This specification is licensed under the
 [FIWARE Open Specification License]
 (http://forge.fiware.org/plugins/mediawiki/wiki/fiware/index.php/FI-WARE_Open_Specification_Legal_Notice_%28essential_patents_license%29).
 
+## FIWARE NGSI information and data model introduction
+
+The FIWARE NGSI (Next Generation Service Interface) API defines 
+
+* a **data model** for context information, based on a simple information model using the notion of *context entities*.
+* a **context data interface** for exchanging information by means of query, subscription, and update operations.
```
Comment (fgalan) [inline doc/apiary/v2/fiware-ngsiv2-reference.apib]: Not sure if from an user perspective is better from him/her to have one integrated interfaces that two sub-interfaces. To be discussed.

File: doc/apiary/v2/fiware-ngsiv2-reference.apib [hunk 6]
```diff
@@ -21,6 +21,47 @@ This specification is licensed under the
...
+* a **context availability interface** for exchanging information on how to obtain context information.
+
+### Context Entities
+
+Context entities, or simply called entities, are the center of gravity in the FIWARE NGSI information model. An entity represents a real-world object like e.g. a sensor, a person, or a room. Each entity is uniquely identified by its **entity id**. 
+
+Furthermore, the type system of FIWARE NGSI enables entities to have an **entity type**. Entity types are semantic types; they are intended to describe the type of real-world object represented by the entity. For example, a context entity with id *sensor-365* could have the type *temperatureSensor*.
+
+### Context Elements
```
Comment (fgalan) [inline doc/apiary/v2/fiware-ngsiv2-reference.apib]: Is is really important the "context element" concept in the API? We have found that use to be difficult to understand by users. On the contrary, entities, attirbutes and metadata are concepts which they find easy to understand. Maybe the best way of defining a context element is as entity + subsect of its attributes. Is that interpretation correct?

**Comment by tobiasjacobs** [comment]
Wow, your feedback is really accurate, thanks!
 
I will mark the "two separate interfaces" issue to be under discussion.

I also agree with Ken not to talk about JSON objects in this general introduction.

I think saying "a context element contains an entity plus a subset of its attributes" is a good idea and consistent with how the distinction was intended.

Stay tuned for another commit of this branch (probably monday)

**Comment by tobiasjacobs** [comment]
by the way, if convenient for you, the apiary-rendered view of this branch can currently be viewed at http://130.206.81.26/ [I have the working copy on that machine]

File: doc/apiary/v2/fiware-ngsiv2-reference.apib (hunk 5 above)
Comment (fgalan) [inline doc/apiary/v2/fiware-ngsiv2-reference.apib]: Fixed in https://github.com/tobiasjacobs/fiware-orion/commit/62031800a91478ad98241e15e0c18a4cd4e00318

(@tobiasjacobs , we use to mark that a given comment has been addressed in a given commit using a "Fixed in X" message as above. That way, specially for long PRs, is easier to trace new changes during the review. Just to let you know ;)).

File: doc/apiary/v2/fiware-ngsiv2-reference.apib (hunk 1 above)
Comment (tobiasjacobs) [inline doc/apiary/v2/fiware-ngsiv2-reference.apib]: fixed in tobiasjacobs@bbea674
(I hope this is the right format?)

File: doc/apiary/v2/fiware-ngsiv2-reference.apib (hunk 5 above)
Comment (tobiasjacobs) [inline doc/apiary/v2/fiware-ngsiv2-reference.apib]: OK, I think I got it :-)

File: doc/apiary/v2/fiware-ngsiv2-reference.apib (hunk 2 above)
Comment (tobiasjacobs) [inline doc/apiary/v2/fiware-ngsiv2-reference.apib]: Fixed by tobiasjacobs@77ed44d

File: doc/apiary/v2/fiware-ngsiv2-reference.apib (hunk 6 above)
Comment (tobiasjacobs) [inline doc/apiary/v2/fiware-ngsiv2-reference.apib]: fixed by tobiasjacobs@77ed44d

I changed the viewpoint accordingly

File: doc/apiary/v2/fiware-ngsiv2-reference.apib (hunk 3 above)
Comment (tobiasjacobs) [inline doc/apiary/v2/fiware-ngsiv2-reference.apib]: Fixed in tobiasjacobs@988709e

File: doc/apiary/v2/fiware-ngsiv2-reference.apib (hunk 4 above)
Comment (tobiasjacobs) [inline doc/apiary/v2/fiware-ngsiv2-reference.apib]: fixed in tobiasjacobs@988709e

File: doc/apiary/v2/fiware-ngsiv2-reference.apib [hunk 7]
```diff
@@ -21,6 +21,48 @@ This specification is licensed under the
...
+The FIWARE NGSI (Next Generation Service Interface) API defines 
+
+* a **data model** for context information, based on a simple information model using the notion of *context entities*
+* a **context data interface** for exchanging information by means of query, subscription, and update operations
+* a **context availability interface** for exchanging information on how to obtain context information
+(whether to separate the two interfaces is currently under discussion).
+
+### Context Entities
+
+Context entities, or simply called entities, are the center of gravity in the FIWARE NGSI information model. An entity represents a real-world object like e.g. a sensor, a person, or a room. Each entity is uniquely identified by its **entity id**.
```
Comment (jmcanterafonseca) [inline doc/apiary/v2/fiware-ngsiv2-reference.apib]: nit: new line needed

File: doc/apiary/v2/fiware-ngsiv2-reference.apib [hunk 8]
```diff
@@ -21,6 +21,48 @@ This specification is licensed under the
...
+* a **data model** for context information, based on a simple information model using the notion of *context entities*
+* a **context data interface** for exchanging information by means of query, subscription, and update operations
+* a **context availability interface** for exchanging information on how to obtain context information
+(whether to separate the two interfaces is currently under discussion).
+
+### Context Entities
+
+Context entities, or simply called entities, are the center of gravity in the FIWARE NGSI information model. An entity represents a real-world object like e.g. a sensor, a person, or a room. Each entity is uniquely identified by its **entity id**. 
+
+Furthermore, the type system of FIWARE NGSI enables entities to have an **entity type**. Entity types are semantic types; they are intended to describe the type of real-world object represented by the entity. For example, a context entity with id *sensor-365* could have the type *temperatureSensor*.
```
Comment (jmcanterafonseca) [inline doc/apiary/v2/fiware-ngsiv2-reference.apib]: new line

**Comment by jmcanterafonseca** [comment]
LGTM provided the new lines are provided that simplifies review and text edition and management

**Comment by tobiasjacobs** [comment]
@jmcanterafonseca 
Not sure if I understand correctly.

Do you mean 
that line breaks 
should be used 
in order to prevent 
overlong text lines
(like done in this 
paragraph)?

File: doc/apiary/v2/fiware-ngsiv2-reference.apib (hunk 6 above)
Comment (fgalan) [inline doc/apiary/v2/fiware-ngsiv2-reference.apib]: I'm still doubtly about it... The concept "context element" is not mentioned in any other piece of the .apib file, so it is seem to be unnecesary to describe and use the API. If you still thinking that we really need it I'd suggest to add the following just after the section tittle:

"(The context element concept is not mentioned in any other piece of the .apib file, so it is seem to be unnecesary to describe and use the API. Thus, this section could be removed. Under discussion)"

**Comment by jmcanterafonseca** [comment]
Yes is what I mean

File: doc/apiary/v2/fiware-ngsiv2-reference.apib (hunk 7 above)
Comment (tobiasjacobs) [inline doc/apiary/v2/fiware-ngsiv2-reference.apib]: done in tobiasjacobs@8f60106

File: doc/apiary/v2/fiware-ngsiv2-reference.apib (hunk 8 above)
Comment (tobiasjacobs) [inline doc/apiary/v2/fiware-ngsiv2-reference.apib]: done in tobiasjacobs@8f60106

File: doc/apiary/v2/fiware-ngsiv2-reference.apib (hunk 6 above)
Comment (tobiasjacobs) [inline doc/apiary/v2/fiware-ngsiv2-reference.apib]: I formulated it in a more neutral and technically more accurate way: In my view what you actually propose is to "merge" the context element and the context entity into a single concept.

done in tobiasjacobs@2da7a11.

**Comment by tobiasjacobs** [comment]
In my view the comments have been addressed and this could be merged now, but let me know if not.

**Comment by kzangeli** [comment]
LGTM

File: doc/apiary/v2/fiware-ngsiv2-reference.apib [hunk 9]
```diff
@@ -21,6 +21,75 @@ This specification is licensed under the
...
+* a **data model** for context information, based on a simple information model using the notion of *context entities*
+* a **context data interface** for exchanging information by means of query, subscription, and update operations
+* a **context availability interface** for exchanging information on how to obtain context information
+(whether to separate the two interfaces is currently under discussion).
+
+### Context Entities
+
+Context entities, or simply called entities, are the center of gravity
+in the FIWARE NGSI information model. An entity represents a
+real-world object like e.g. a sensor, a person, or a room. Each entity
```
Comment (fgalan) [inline doc/apiary/v2/fiware-ngsiv2-reference.apib]: Actually, an entity can represent things of logical nature that strictly speking are not part of the real-works, e.g. an "alarm" entity within a trouble ticketing system.

However, this is a very minor comment, and we can adjust the description in a next PR.

NTC

**Comment by fgalan** [comment]
LGTM

**Pull request merged**